### PR TITLE
Add compression and better error handling

### PR DESCRIPTION
**Pull request by juliens**
### What does this PR do?
- add compression if client send the compression header and backend valid compression
- handle close message correctly and log only real error

### More
- [x] tests

### CODE REVIEW

File: forward/fwd.go [hunk 1]
```diff
@@ -331,29 +346,45 @@ func (f *httpForwarder) serveWebSocket(w http.ResponseWriter, req *http.Request,
 	defer underlyingConn.Close()
 	defer targetConn.Close()
 
-	errc := make(chan error, 2)
-
-	replicateWebsocketConn := func(dst, src *websocket.Conn, dstName, srcName string) {
-		var err error
+	errclient := make(chan error)
+	errbackend := make(chan error)
```
Comment (ldez) [inline forward/fwd.go]: /errbackend/errBackend/
/errclient/errClient/

File: forward/fwd_websocket_test.go [hunk 2]
```diff
@@ -51,6 +52,106 @@ func (s *FwdSuite) TestWebSocketEcho(c *C) {
...
+	conn, resp, err := dialer.Dial(webSocketURL, headers)
+	if err != nil {
+		c.Errorf("Error [%s] during Dial with response: %+v", err, resp)
+		return
+	}
+	conn.WriteMessage(gorillawebsocket.TextMessage, []byte("OK"))
+	c.Log(conn.ReadMessage())
+	isCompress := strings.Contains(resp.Header.Get("Sec-Websocket-Extensions"), "permessage-deflate")
+	c.Assert(isCompress, Equals, true)
+
```
Comment (ldez) [inline forward/fwd_websocket_test.go]: could you remove this line and add some before ?

File: forward/fwd_websocket_test.go [hunk 3]
```diff
@@ -51,6 +52,106 @@ func (s *FwdSuite) TestWebSocketEcho(c *C) {
...
+	conn, resp, err := dialer.Dial(webSocketURL, headers)
+	if err != nil {
+		c.Errorf("Error [%s] during Dial with response: %+v", err, resp)
+		return
+	}
+	conn.WriteMessage(gorillawebsocket.TextMessage, []byte("OK"))
+	c.Log(conn.ReadMessage())
+	isCompress := strings.Contains(resp.Header.Get("Sec-Websocket-Extensions"), "permessage-deflate")
+	c.Assert(isCompress, Equals, false)
+
```
Comment (ldez) [inline forward/fwd_websocket_test.go]: could you remove this line and add some before ?

**Review by ldez — CHANGES_REQUESTED** [review]

**Review by ldez — APPROVED** [review]
LGTM

File: forward/fwd.go [hunk 4]
```diff
@@ -281,6 +282,11 @@ func (f *httpForwarder) serveWebSocket(w http.ResponseWriter, req *http.Request,
 	outReq := f.copyWebSocketRequest(req)
 
 	dialer := websocket.DefaultDialer
+
+	if strings.Contains(req.Header.Get("Sec-Websocket-Extensions"), "permessage-deflate") {
+		dialer.EnableCompression = true
+	}
```
Comment (ldez) [inline forward/fwd.go]:
```go
dialer.EnableCompression = strings.Contains(req.Header.Get("Sec-Websocket-Extensions"), "permessage-deflate")
```

File: forward/fwd.go [hunk 5]
```diff
@@ -322,7 +328,12 @@ func (f *httpForwarder) serveWebSocket(w http.ResponseWriter, req *http.Request,
 		return true
 	}}
 
+	if strings.Contains(resp.Header.Get("Sec-Websocket-Extensions"), "permessage-deflate") {
+		upgrader.EnableCompression = true
+	}
```
Comment (ldez) [inline forward/fwd.go]:
```go
upgrader.EnableCompression = strings.Contains(resp.Header.Get("Sec-Websocket-Extensions"), "permessage-deflate")
```

**Review by ldez — CHANGES_REQUESTED** [review]

**Review by ldez — APPROVED** [review]
LGTM

**Review by mmatur — APPROVED** [review]
LGTM

**Review by nmengin — APPROVED** [review]

**Pull request merged**